src/modm/driver/adc/ads816x_impl.hpp

```diff
@@ -0,0 +1,134 @@
+// coding: utf-8
+// ----------------------------------------------------------------------------
+/*
+ * Copyright (c) 2021, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ADS816X_HPP
+#	error "Don't include this file directly! Use 'ads816x.hpp' instead."
+#endif
+
+#include <bit>
+
+namespace modm
+{
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<void>
+Ads816x<SpiMaster, Cs>::initialize(Mode mode)
+{
+	using namespace std::chrono_literals;
+	RF_BEGIN();
+
+	// Unlock register access
+	RF_CALL(registerAccess(Command::Write, Register::REG_ACCESS, 0b1010'1010));
+
+	timeout.restart(4us);
+	RF_WAIT_UNTIL(timeout.isExpired());
+
+	// Set mode
+	RF_CALL(registerAccess(Command::Write, Register::DEVICE_CFG, uint8_t(mode)));
+
+	RF_END();
+}
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<uint16_t>
+Ads816x<SpiMaster, Cs>::manualModeConversion(uint8_t nextChannel)
+{
+	RF_BEGIN();
+
+	RF_WAIT_UNTIL(this->acquireMaster());
+	Cs::reset();
+
+	buffer[0] = uint8_t(Command::Write) | (uint16_t(Register::CHANNEL_ID) >> 8);
+	buffer[1] = uint16_t(Register::CHANNEL_ID);
+	buffer[2] = (nextChannel & 0b0000'0111);
+
+	RF_CALL(SpiMaster::transfer(buffer, buffer2, 3));
+
+	if (this->releaseMaster()) {
+		Cs::set();
+	}
+
+	RF_END_RETURN( static_cast<uint16_t>((static_cast<uint16_t>(buffer2[0]) << 8) | buffer2[1]) );
+}
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<void>
+Ads816x<SpiMaster, Cs>::autoSequenceConversion(uint8_t channelsBitmask, std::array<uint16_t, 8>& result)
```

Comment:
Wouldn't it be better to use `std::span` instead of `std::array&`? There is a choice to be made with span if it should be a fixed size buffer (`std::span<uint16_t, 8>`) or we make it dynamically sized to accomodate for the use-case of sampling less than 8 channels  (`std::span<uint16_t>`). The dynamic size would imply a runtime check for the size.

Reply:
Good idea